src/Stratis.Bitcoin.Features.Consensus.Tests/TimeSyncBehaviorAdjustedTimeTest.cs

```diff
@@ -0,0 +1,96 @@
+
+        /// <summary>
+        /// This provides the adjusted time ready for assertion. 
+        /// </summary>
+        private IDateTimeProvider dateTimeProvider;
+        
+        /// <summary>
+        /// This is the result of the operation under test. 
+        /// </summary>
+        private TimeSpan offset;
+
+        /// <summary>
+        /// Checks that <see cref="TimeSyncBehaviorState.AddTimeData(IPAddress, TimeSpan, bool)"/>
+        /// uses outbound samples as a priority over inbound as they are less likely to be malicious.
+        /// This means that even when there are many inbound connections, only the first ones are considered until there are enough outbound nodes. 
+        /// Eg only 9 inbound per 1 outbound.
+        /// </summary>
+        [Fact]
+        public void AddTimeData_WithLargeSampleSetOfInboundTimeManipulatorsAndLowSampleSetOfOutbound_GetsOverridenByOutboundSamples()
+        {
+            this.Given_an_empty_time_sync_behaviour_state();
+            this.Given_40_inbound_samples_with_offset_of(10);
+            this.Given_1_outbound_sample_with_offset_of(20);
+            this.When_calculating_time_adjust_offset();
+            this.Then_adjusted_time_offset_should_be(20);
+        }
+
+        /// <summary>
+        /// Checks that <see cref="TimeSyncBehaviorState.AddTimeData(IPAddress, TimeSpan, bool)"/>
+        /// uses system time without an offset when there are no outbound samples.
+        /// </summary>
+        [Fact]
+        public void AddTimeData_WithLargeSampleSetOfInboundTimeManipulatorsAndZeroOutbound_SticksWithTheSystemTime()
+        {
+            this.Given_an_empty_time_sync_behaviour_state();
+            this.Given_40_inbound_samples_with_offset_of(10);
+            this.When_calculating_time_adjust_offset();
+            this.Then_adjusted_time_offset_should_be(0);
+        }
+
+        private void Given_an_empty_time_sync_behaviour_state()
+        {
+            this.dateTimeProvider = new DateTimeProvider();
+            this.timesyncBehaviourState = new TimeSyncBehaviorState(
+                this.dateTimeProvider, new NodeLifetime(),
+                new AsyncLoopFactory(new LoggerFactory()), 
+                new LoggerFactory());
+        }
+
+        private void Given_1_outbound_sample_with_offset_of(int offsetSeconds)
+        {
+            this.timesyncBehaviourState.AddTimeData(IPAddress.Parse("2.2.2.2"), TimeSpan.FromSeconds(offsetSeconds), isInboundConnection: false);
+        }
+
+        private void Given_40_inbound_samples_with_offset_of(int offset)
+        {
+            for (int i = 1; i <= 40; i++)
+            {
+                this.timesyncBehaviourState.AddTimeData(IPAddress.Parse("1.2.3." + i), TimeSpan.FromSeconds(offset), isInboundConnection: true);
+            }
+        }
+
+        private void When_calculating_time_adjust_offset()
+        {
+            var adjustedTime = this.dateTimeProvider.GetAdjustedTime();
```

Comment:
unvar here, on the next line and line 91

Reply:
var is very clear here

Reply:
is it? so is time given as DateTime or UNIX timestamp here? without looking anywhere, i can't remember myself. and if unix timestamp, is it int, uint or long?

Reply:
as a rule of thumb - if you have `new` on the right side, use `var`, otherwise don't unless the type is unknown (anonymous)

Reply:
looking at the line of code in isolation like the above, yes, you might not be able to tell the type. So you are factually correct. However, perhaps you aren't considering why you need to know the type at that moment in time?  It doesn't degrade from your understanding and editing ability by not knowing the type from that one line of code.

However when you put that line of code back in context it came from, it is very clear that it is a DateTime just like it is clear that the line below is a DateTime too.

>var adjustedTime = this.dateTimeProvider.GetAdjustedTime();
>var now = this.dateTimeProvider.GetUtcNow();
>this.offset = adjustedTime - now;

I would concede that there is an improvement here if we call offset an  adjustedOffsetTimespan. After that it is 100% clear.

therefore reaffirming that it is of zero additional value to explicitly name the type. If you had to edit this section of code, it would be immediately obvious it was a DateTime. 

I still accept that if it is the policy then I will obviously follow that rule now and make my case for changing it :)

Reply:
The reason for specifying the type is that when you use `var` and read the code you kinda almost always have to move your pointer to that variable and look at the VS hint that shows you the actual type. 
So not using var when it's not obvious saves some time while reading the code and makes it more clear.

Reply:
in fact here, these two variables add nothing, so I will inline them

Reply:
Actually your long description is exactly giving a good reason against using Var here. You just explained that your brain need to do another operation to get that information. And that is exactly what we try to avoid

Reply:
The same principle justifies rule number 23 encoding style – using parenthesis that the compiler itself does not need, but it helps human brain to save couple of XYZseconds